File: DSA/Java/src/dsa/lib/algo/sort/CountingSort.java
package dsa.lib.algo.sort;

import dsa.lib.utils.Arrays;

public class CountingSort {

    private int[] nums;
    private final int size;

    int min, max;
    int[] counts;

    public CountingSort(int[] nums) {
        this.nums = nums;
        this.size = nums.length;
    }

    public void sort() {
        if (this.size < 2) return;

        initFields();
        storeCounts();
        accumulateCounts();

        int[] result = new int[this.size];
        generateResult(result);
        Arrays.copy(result, 0, this.nums, 0, result.length);
    }

    private void initFields() {
        initMinMax();
        initCounts();
    }

    public void initMinMax() {
        if (this.size == 0) return;

        min = max = this.nums[0];
        for (final int num : nums) {
            if (num < min) min = num;
            else if (num > max) max = num;
        }
    }

    public void initCounts() {
        counts = new int[max - min + 1];
        Arrays.fill(counts, 0);
    }

    public void storeCounts() {
        for (final int num : this.nums) ++counts[countsIndexOf(num)];
    }

    private int countsIndexOf(int num) {
        return num - min;
    }

    public void accumulateCounts() {
        for (int i = 1; i < counts.length; i++) counts[i] += counts[i - 1];
    }

    private void generateResult(int[] result) {
        for (int i = this.size - 1; i >= 0; --i) {
            int num = this.nums[i];
            final int ci = countsIndexOf(num);
            result[counts[ci] - 1] = num;
            --counts[ci];
        }
    }
}

File: DSA/Python/src/dsa/lib/algo/sort/counting_sort.py
from typing import List

from dsa.lib import utils


class CountingSort:

    def __init__(self, nums: List[int]) -> None:
        self.nums = nums
        self.size = len(nums)

        self.min = None
        self.max = None
        self.counts = None

    def sort(self) -> None:
        if self.size < 2: return

        self.init_fields()
        self.store_counts()
        self.accumulate_counts()

        result = [0] * self.size
        self.generate_result(result)
        utils.array_copy(result, 0, self.nums, 0, len(result))

    def init_fields(self):
        self.init_minMax()
        self.init_counts()

    def init_minMax(self):
        if not self.size: return

        self.min = self.max = self.nums[0]
        for num in self.nums:
            if num > self.max:
                self.max = num
            elif num < self.min:
                self.min = num

    def init_counts(self):
        self.counts = [0] * (self.max - self.min + 1)

    def store_counts(self):
        for num in self.nums: self.counts[self.countsIndexOf(num)] += 1

    def countsIndexOf(self, num: int) -> int:
        return num - self.min

    def accumulate_counts(self):
        for i in range(1, len(self.counts)): self.counts[i] += self.counts[i - 1]

    def generate_result(self, result: List[int]) -> None:
        for i in range(self.size - 1, -1, -1):
            num = self.nums[i]
            ci = self.countsIndexOf(num)
            result[(self.counts[ci] - 1)] = num
            self.counts[ci] -= 1
